jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,622 @@
+
+        if (id == null)
+            return _rejectHandler.handle(request, response, callback);
+
+        // Obtain a tracker, creating a new one if necessary.  Trackers are removed if CyclicTimeouts#onExpired returns true
+        Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
+
+        // If we are not over-limit then handle normally
+        if (!tracker.isRateExceededBySample(request.getBeginNanoTime()))
+            return nextHandler(request, response, callback);
+
+        // Otherwise reject the request
+        return _rejectHandler.handle(request, response, callback);
+    }
+
+    Tracker newTracker(String id)
+    {
+        return new Tracker(id, _rateControlFactory.newRateControl());
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _cyclicTimeouts = new CyclicTimeouts<>(getServer().getScheduler())
+        {
+            @Override
+            protected Iterator<Tracker> iterator()
+            {
+                return _trackers.values().iterator();
+            }
+
+            @Override
+            protected boolean onExpired(Tracker tracker)
+            {
+                return tracker.isIdle(NanoTime.now());
+            }
+        };
+        addBean(_cyclicTimeouts);
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(_cyclicTimeouts);
+        _cyclicTimeouts.destroy();
+        _cyclicTimeouts = null;
+    }
+
+    /**
+     * A RateTracker is associated with a connection, and stores request rate data.
+     */
+    class Tracker implements CyclicTimeouts.Expirable
+    {
+        private final AutoLock _lock = new AutoLock();
+        private final String _id;
+        private final RateControl _rateControl;
+        private final Duration _idleCheck;
+        private long _expireAt;
+
+        Tracker(String id, RateControl rateControl)
+        {
+            this(id, rateControl, null);
+        }
+
+        Tracker(String id, RateControl rateControl, Duration idleCheck)
+        {
+            _id = id;
+            _rateControl = rateControl;
+            _idleCheck = idleCheck == null ? Duration.ofSeconds(2) : idleCheck;
+            _expireAt = NanoTime.now() + _idleCheck.toNanos();
+        }
+
+        public String getId()
+        {
+            return _id;
+        }
+
+        RateControl getRateControl()
+        {
+            return _rateControl;
+        }
+
+        public boolean isRateExceededBySample(long now)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                CyclicTimeouts<Tracker> cyclicTimeouts = _cyclicTimeouts;
+                if (cyclicTimeouts != null)
+                {
+                    // schedule a check to remove this tracker if idle
+                    _expireAt = now + _idleCheck.toNanos();
+                    cyclicTimeouts.schedule(this);
+                }
```

Comment:
I don't understand this. This class is already `Expirable`, so it just needs to update the `_expireAt`.

Reply:
ummm no! You still need to actually schedule it.

Reply:
You just schedule it at creation.

Reply:
ummmm no!  The whole point of a cyclic timeout is that it is optimised to be cancelled.  If we are getting requests, we don't need to expire the timeout, we want to set it again into the future. We only want to expire if we have not received any requests for a period of time.

In fact, I think I'm missing a schedule for when it expires, but is not idle!

Reply:
> we don't need to expire the timeout, we want to set it again into the future

That's exactly what CyclicTimeouts does.

Reply:
> > we don't need to expire the timeout, we want to set it again into the future
> 
> That's exactly what CyclicTimeouts does.

Then the contract is strange and we have gotten it "wrong" all over the code base:
 + https://github.com/jetty/jetty.project/blob/jetty-12.0.13/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java#L111
 + https://github.com/jetty/jetty.project/blob/jetty-12.0.13/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpDestination.java#L319
 + https://github.com/jetty/jetty.project/blob/jetty-12.0.13/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java#L352

I can see a few places where `schedule` is called from just a `setIdleTimeout` method, so it is relying on the ability to just mutate the return of `getExpireNanoTime`, but then the `schedule` method is named incorrectly.  It should just be called `add`.      This is also contradictory to the behaviour in `CyclicTimeout`, where a call to `schedule` is needed after a timeout has expired.

So I do not think it wrong to call `schedule`, but it may be unnecessary.    The contract/naming of `CyclicTimeouts` needs to be improved so that it is used consistently.

Reply:
@gregw see https://github.com/jetty/jetty.project/issues/12308#issuecomment-2374252289.